### PR TITLE
Save LogCat output if unit tests fail.

### PR DESCRIPTION
**Pull request by cmelchior**
If unit tests crashes with native exceptions we currently have no way of knowing why.
This change means that we now save the logcat output from in case of unit test failures.

A local test on my phone indicates that a full logcat file is ~8MB in size and 400Kb when zipped.

### CODE REVIEW

**Comment by cmelchior** [comment]
CI: Retest this please

File: Jenkinsfile [hunk 1]
```diff
@@ -87,6 +92,27 @@
     }
 }
 
+def startLogCatCollector() {
+    sh '''rm -f logcat.txt
```
Comment (emanuelez) [inline Jenkinsfile]: this is not needed since a `git clean -xfd` is ran after checkout

File: Jenkinsfile [hunk 2]
```diff
@@ -87,6 +92,27 @@
...
+    adb logcat -c
+    adb logcat > "logcat.txt" &
+    '''
+}
+
+def stopLogCatCollector(String pid, boolean archiveLog) {
+    sh '''jobs
+       kill %1'
```
Comment (emanuelez) [inline Jenkinsfile]: I suppose `jobs` is just for debugging purposes, right?
if you use double quotes you will be able to use the `pid` argument

File: Jenkinsfile [hunk 3]
```diff
@@ -87,6 +92,27 @@
...
+       '''
+    if (archiveLog) {
+        zip([
+            'zipFile': 'logcat.zip',
+            'archive': true,
```
Comment (emanuelez) [inline Jenkinsfile]: true -> archivelog

Comment (cmelchior) [inline Jenkinsfile]: You mean `'archive' : 'archivelog',` ? 
This https://jenkins.io/doc/pipeline/steps/pipeline-utility-steps/ says it is a boolean?

Comment (beeender) [inline Jenkinsfile]: if it take `archiveLog` then `if (archiveLog)` is not needed i think?

**Comment by beeender** [comment]
can we get a download link to it :8ball:

File: Jenkinsfile (hunk 3 above)
Comment (emanuelez) [inline Jenkinsfile]: oops... missed that. sorry

File: Jenkinsfile [hunk 4]
```diff
@@ -87,6 +92,26 @@
     }
 }
 
+def startLogCatCollector() {
+    sh '''adb logcat -c
+    adb logcat > "logcat.txt" &
+    '''
+}
+
+def stopLogCatCollector(boolean archiveLog) {
```
Comment (emanuelez) [inline Jenkinsfile]: I think we should use a PID. this version is fragile as other poeple might be logged into the host and jobs might return unexpected results

File: Jenkinsfile (hunk 3 above)
Comment (emanuelez) [inline Jenkinsfile]: no need to zip if everything passes. We should keep it as it is

File: Jenkinsfile (hunk 4 above)
Comment (cmelchior) [inline Jenkinsfile]: For some reason I didn't get the PID in the previous version. No idea why?

File: Jenkinsfile [hunk 5]
```diff
@@ -87,6 +93,25 @@
...
 
+def String startLogCatCollector() {
+    sh '''adb logcat -c
+    `adb logcat > "logcat.txt" &` & echo $1 > pid
+    '''
+    return readFile("pid").trim()
+}
+
+def stopLogCatCollector(String backgroundPid, boolean archiveLog) {
+    sh 'kill $backgroundPid'
```
Comment (emanuelez) [inline Jenkinsfile]: I hope SIGTERM will be enough in all cases. But I would of course prefer not to use SIGKILL.

File: Jenkinsfile [hunk 6]
```diff
@@ -87,6 +93,25 @@
     }
 }
 
+def String startLogCatCollector() {
```
Comment (emanuelez) [inline Jenkinsfile]: it's not very Groovysh to specify the return type, but I guess it doesn't hurt

File: Jenkinsfile [hunk 7]
```diff
@@ -87,6 +93,26 @@
...
+    return readFile("pid").trim()
+}
+
+def stopLogCatCollector(String backgroundPid, boolean archiveLog) {
+    sh "kill ${backgroundPid}"
+    if (archiveLog) {
+        zip([
+            'zipFile': 'logcat.zip',
+            'archive': true,
+            'glob' : '**/logcat.txt'
```
Comment (emanuelez) [inline Jenkinsfile]: such a wildcard can potentially make looking for the file very slow. We know where the file is so let's use the proper path

Comment (cmelchior) [inline Jenkinsfile]: Yeah, but it didn't find the file for some reason when using `realm/logcat.txt` ... 😢

**Comment by emanuelez** [comment]
See? That was easy! :)

**Comment by cmelchior** [comment]
Wrong branch. Moving this to `releases`

**Pull request closed without merging**